fw/http.c

```diff
@@ -4990,88 +4968,63 @@ tfw_h2_error_resp(TfwHttpReq *req, int status, bool reply, bool attack,
 	 * and GOAWAY frame should be sent (RFC 7540 section 6.8) after
 	 * error response.
 	 */
-	if (reply) {
-		tfw_h2_send_err_resp(req, status, 0);
-		if (attack)
-			tfw_h2_conn_terminate_close(ctx, HTTP2_ECODE_PROTO,
-						    !on_req_recv_event);
-		return;
+	tfw_h2_send_err_resp(req, status, stream->id);
+	if (attack) {
+		tfw_h2_conn_terminate_close(ctx, HTTP2_ECODE_PROTO,
+					    !on_req_recv_event);
+	} else {
+		if (STREAM_SEND_PROCESS(stream, HTTP2_RST_STREAM, 0))
+			return TFW_BAD;
+		tfw_h2_send_rst_stream(ctx, stream->id, HTTP2_ECODE_PROTO);
 	}
-
-	 /*
-	  * If no reply is needed and this is not the attack case, the
-	  * connection needn't to be closed, and we should indicate to
-	  * remote peer (via RST_STREAM frame), that the stream has entered
-	  * into closed state (RFC 7540 section 6.4).
-	  */
-	stream_id = tfw_h2_stream_id_send(req, HTTP2_RST_STREAM, 0);
-	if (stream_id && !attack)
-		tfw_h2_send_rst_stream(ctx, stream_id, HTTP2_ECODE_CANCEL);
+	goto out;
 
 skip_stream:
 	if (attack) {
-		if (reply) {
-			tfw_h2_conn_terminate_close(ctx, HTTP2_ECODE_PROTO,
-						    !on_req_recv_event);
-		} else {
-			do_access_log_req(req, 403, 0);
-			if (!on_req_recv_event)
-				tfw_connection_close(req->conn, true);
-		}
+		tfw_h2_conn_terminate_close(ctx, HTTP2_ECODE_PROTO,
+					    !on_req_recv_event);
 	}
 
+free_req:
+	do_access_log_req(req, status, 0);
 	tfw_http_conn_msg_free((TfwHttpMsg *)req);
+out:
+	return reply ? (attack ? TFW_BAD : TFW_DROP) : TFW_BLOCK;
 }
 
-static void
+static int
 tfw_h1_error_resp(TfwHttpReq *req, int status, bool reply, bool attack,
 		  bool on_req_recv_event)
 {
+	int r = TFW_DROP;
 	TfwCliConn *cli_conn = (TfwCliConn *)req->conn;
 
 	/* The client connection is to be closed with the last resp sent. */
 	reply &= !test_bit(TFW_HTTP_B_REQ_DROP, req->flags);
```

Comment:
Please pay attention on this place. Is it correct?

Reply:
As I understand we should rely on `connection: close/keep-alive` header.

Reply:
I think that the behavior should be same for HTTP1 and HTTP2. In case when no reply is needed we should close connection with TCP RST without any response. If attack occurred and we reply is true we should send reply and close connection with TCP FIN, since we can't send reply if we close connection with TCP RST.   For error we should close connection if we can't work with this client (parser errors) and  continue in other cases.